File: ext/cpp_polygon_finder/PolygonFinder/src/polygon/finder/NodeCluster.cpp
/*
 * NodeCluster.cpp
 *
 * Copyright (c) 2025-2026 Emanuele Cesaroni
 *
 * Licensed under the GNU Affero General Public License v3 (AGPLv3).
 * See the LICENSE file in this directory for the full license text.
 */

#include <iostream>
#include <list>
#include <algorithm>
#include <map>
#include <string>
#include <vector>
#include <utility>
#include <deque>
#include "NodeCluster.h"
#include "Node.h"
#include "RectBounds.h"
#include "../reducers/UniqReducer.h"
#include "../reducers/LinearReducer.h"
#include "../reducers/VisvalingamReducer.h"

NodeCluster::NodeCluster(int h, int w, pf_Options *options) {
  versus_inverter[Node::O] = Node::A;
  versus_inverter[Node::A] = Node::O;
  this->options = options;
  this->height = h;
  this->width = w;
  this->nodes = 0;

  this->vert_nodes.resize(h);
  this->root_nodes = this->lists.add_list();
  this->inner_plot = this->lists.add_list();
  this->inner_new = this->lists.add_list();
}

NodeCluster::~NodeCluster() {
}

void NodeCluster::compress_coords(std::list<Polygon>& polygons, pf_Options options) {
  if (!(options.compress_linear || options.compress_uniq || options.compress_visvalingam)) return;

  auto compress_sequence = [&](std::vector<Point*>& points_vec) {
    if (points_vec.empty()) return;

    if (options.compress_uniq) {
      UniqReducer uniq_reducer(points_vec);
      uniq_reducer.reduce();
    }
    if (options.compress_linear) {
      LinearReducer linear_reducer(points_vec);
      linear_reducer.reduce();
    }
    if (options.compress_visvalingam) {
      VisvalingamReducer vis_reducer(points_vec, options.compress_visvalingam_tolerance);
      vis_reducer.reduce();
    }
  };
  for (auto &poly : polygons) {
    compress_sequence(poly.outer);
    for (auto &inner_seq : poly.inner) {
        compress_sequence(inner_seq);
    }
  }
}

void NodeCluster::build_tangs_sequence() {
  for (auto& line : vert_nodes) {
    for (Node& node : line) {
      node.precalc_tangs_sequences(*this);
    }
  }
}

Node* NodeCluster::add_node(int min_x, int max_x, int y, char name, int offset) {
  vert_nodes[y].emplace_back(min_x, max_x, y, this, name);

  Node& node = vert_nodes[y].back();
  node.abs_x_index = vert_nodes[y].size() - 1;
  this->nodes++;

  root_nodes->push_back(&node);

  if (y > 0) {
    std::deque<Node>& up_nodes = vert_nodes[y - 1];
    if (!up_nodes.empty()) {
      auto it = std::lower_bound(up_nodes.begin(), up_nodes.end(), node.min_x,
        [&](const Node& a, int val) {
          return ((a.max_x + offset) < val);
        });

      while (it != up_nodes.end()) {
        if ((it->min_x - offset) > node.max_x) break;
        int current_index = std::distance(up_nodes.begin(), it);
        node.add_intersection(*it, current_index);
        it->add_intersection(node, node.abs_x_index);
        ++it;
      }
    }
  }
  return &node;
}

void NodeCluster::plot(int versus) {
  int inner_v = versus_inverter[versus];
  int index_order = 0;
  Node *next_node;

  while (root_nodes->size() > 0)
  { Node *root_node = reinterpret_cast<Node*>(root_nodes->shift());
    root_node->outer_index = index_order;
    this->plot_sequence.clear();
    this->plot_sequence.push_back(root_node);
    Polygon poly;

    if ((root_node)->tangs_sequence.size() > 0)  // front() or back() on empty list is undefined
    { versus == Node::A ?
        next_node = root_node->get_tangent_node_by_virtual_index(root_node->tangs_sequence.back()) :
        next_node = root_node->get_tangent_node_by_virtual_index(root_node->tangs_sequence.front());

      if (next_node != nullptr)
      { Point* p = next_node->coords_entering_to(root_node, versus_inverter[versus], Node::OUTER);
        poly.outer.push_back(p);
        poly.bounds.expand(p->x, p->y);
      }
      if ((this->nodes > 0) && (next_node != nullptr))
      { plot_node(poly.outer, next_node, root_node, versus, poly.bounds);
      }
      this->sequences.push_back(std::move(this->plot_sequence));
      this->plot_sequence.clear();

      if (poly.outer.size() > 2)
      { this->polygons.push_back(poly);
      }

      int index_inner = 0;
      while (inner_plot->size() > 0)
      { this->plot_sequence.clear();
        std::vector<Point*> inner_sequence;
        std::list<Node*>::iterator first_i;
        Node *first = nullptr;

        Listable *act = inner_plot->first();
        do
        { if ((reinterpret_cast<Node*>(act))->tangs_count <= 2)
          { first = reinterpret_cast<Node*>(act);
            break;
          }
        }while((act = act->data_pointer[inner_plot->get_id()].next) != nullptr);

        if (first == nullptr) first = reinterpret_cast<Node*>(inner_plot->first());

        plot_sequence.push_back(first);
        inner_plot->remove(first);
        root_nodes->remove(first);

        first->inner_index = index_inner;

        Node* next_node = nullptr;

        if (first->get_trackmax()) {
          if (inner_v == Node::A) {
            if (first->upper_end >= 0) next_node = &this->vert_nodes[first->y + Node::T_UP][first->upper_start];
          } else {
            if (first->lower_end >= 0) next_node = &this->vert_nodes[first->y + Node::T_DOWN][first->lower_start];
          }
        } else {
          if (inner_v == Node::A) {
            if (first->lower_end >= 0) next_node = &this->vert_nodes[first->y + Node::T_DOWN][first->lower_end];
          } else {
            if (first->upper_end >= 0) next_node = &this->vert_nodes[first->y + Node::T_UP][first->upper_end];
          }
        }

        if (next_node != nullptr)
        { inner_sequence.push_back(next_node->coords_entering_to(first, inner_v, Node::INNER));
          plot_inner_node(inner_sequence, next_node, inner_v, first, root_node);
        }
        this->polygons.back().inner.push_back(inner_sequence);
        this->inner_plot->grab(this->inner_new);
        index_inner++;
      }
    } else {
      this->sequences.push_back(std::move(this->plot_sequence));
      this->plot_sequence.clear();
    }
    // tree
    if (this->options->treemap)
    { this->treemap.push_back(versus == Node::A ? this->test_in_hole_a(root_node) : this->test_in_hole_o(root_node));
    }
    index_order++;
  }
}

std::pair<int, int> NodeCluster::test_in_hole_a(Node* node)
{ if (node->outer_index > 0)
  { int start_left = node->abs_x_index - 1;
    do {
      Node* prev = &this->vert_nodes[node->y][start_left];
      int cindex = prev->outer_index;

      if ((cindex < node->outer_index) && (prev->track & Node::IMAX))
      { unsigned int start_right = node->abs_x_index;
        unsigned int line_size = this->vert_nodes[node->y].size();
        while (++start_right != line_size)
        { Node* tnext = &this->vert_nodes[node->y][start_right];
          if (tnext->outer_index == cindex) {
            if (tnext->track & Node::IMIN) return {cindex, prev->inner_index};
            else                           return {-1, -1};
          }
        }
      }
    } while (--start_left >= 0);
  }
  return {-1, -1};
}

std::pair<int, int> NodeCluster::test_in_hole_o(Node* node)
{ auto& line = this->vert_nodes[node->y];
  const unsigned int line_size = line.size();
  if (node->outer_index == 0 || &line.back() == node) return {-1, -1};

  unsigned int start_left = node->abs_x_index + 1;
  do {
    Node* prev = &line[start_left];
    int cindex = prev->outer_index;

    if (cindex < node->outer_index && (prev->track & Node::IMIN))
    { int start_right = node->abs_x_index;
      while (--start_right >= 0) {
        Node* tnext = &line[start_right];
        if (tnext->outer_index == cindex) {
          if (tnext->track & Node::IMAX) return {cindex, prev->inner_index};
          else                           return {-1, -1};
        }
      }
    }
  } while (++start_left != line_size);
  return {-1, -1};
}

void NodeCluster::plot_inner_node(std::vector<Point*>& sequence_coords, Node *node, int versus, Node *stop_at, Node *start_node) {
  Node *current_node = node;

  while (current_node != nullptr) {
    current_node->outer_index = start_node->outer_index;
    current_node->inner_index = stop_at->inner_index;
    root_nodes->remove(current_node);
    inner_plot->remove(current_node);

    Node *last_node = plot_sequence.back();
    Node *next_node = current_node->my_next_inner(last_node, versus);

    plot_sequence.push_back(current_node);

    bool plot = true;
    if (next_node->y == last_node->y) {
      Node *n;
      int virtual_index = (versus == Node::A ?
        current_node->tangs_sequence.front() :
        current_node->tangs_sequence.back());
      n = current_node->get_tangent_node_by_virtual_index(virtual_index);
      plot = (n == next_node);
    }
    if (plot) {
      sequence_coords.push_back(last_node->coords_entering_to(current_node, versus_inverter[versus], Node::INNER));
      if (current_node != start_node) {
        if (last_node->y == next_node->y) {
          sequence_coords.push_back(next_node->coords_entering_to(current_node, versus, Node::INNER));
        }
      }
    }
    if (current_node->track_uncomplete()) {
      this->inner_new->push_back(current_node);
    } else {
      inner_new->remove(current_node);
    }
    if (next_node == stop_at) break;
    current_node = next_node;
  }
}

void NodeCluster::plot_node(std::vector<Point*>& sequence_coords, Node *node, Node *start_node, int versus, RectBounds& bounds) {
  Node *current_node = node;

  while (current_node != nullptr) {
    root_nodes->remove(current_node);
    current_node->outer_index = start_node->outer_index;

    Node *last_node = plot_sequence.back();
    Node *next_node = current_node->my_next_outer(last_node, versus);

    plot_sequence.push_back(current_node);

    bool plot = true;
    if (next_node->y == last_node->y) {
      int virtual_index = (versus == Node::A ?
          current_node->tangs_sequence.back() :
          current_node->tangs_sequence.front());
      Node *n = current_node->get_tangent_node_by_virtual_index(virtual_index);
      plot = (n == next_node);
    }
    if (plot) {
      Point* p = last_node->coords_entering_to(current_node, versus, Node::OUTER);
      sequence_coords.push_back(p);
      bounds.expand(p->x, p->y);
      if (current_node != start_node) {
        inner_plot->contains(current_node) ? inner_plot->remove(current_node) : inner_plot->push_back(current_node);
        if (last_node->y == next_node->y) {
          Point* p1 = next_node->coords_entering_to(current_node, versus_inverter[versus], Node::OUTER);
          sequence_coords.push_back(p1);
          bounds.expand(p1->x, p1->y);
          inner_plot->contains(current_node) ? inner_plot->remove(current_node) : inner_plot->push_back(current_node);
        }
      }
    }
    if (current_node == start_node) {
      if (current_node->track_complete()) break;
    }
    current_node = next_node;
  }
}

File: lib/contrek/finder/node_cluster.rb
# frozen_string_literal: true

module Contrek
  module Finder
    class NodeCluster
      attr_reader :root_nodes, :sequences, :polygons, :lists, :treemap, :vert_nodes, :options
      VERSUS_INVERTER = {a: :o, o: :a}

      def initialize(h, options)
        @options = options
        @vert_nodes = Array.new(h) { [] }
        @sequences = []
        @polygons = []
        @treemap = []
        @nodes = 0
        @lists = Contrek::Finder::Lists.new
        @root_nodes = @lists.add_list
        @inner_plot = @lists.add_list
        @inner_new = @lists.add_list
      end

      def path_sequences
        @polygons.compact.each do |polygon|
          yield polygon[:outer]
          polygon[:inner].each do |sequence|
            yield sequence
          end
        end
      end

      def compress_coords
        path_sequences do |seq|
          Contrek::Reducers::UniqReducer.new(points: seq).reduce! if @options[:compress].has_key?(:uniq)
          Contrek::Reducers::LinearReducer.new(points: seq, options: @options[:compress][:linear]).reduce! if @options[:compress].has_key?(:linear)
          Contrek::Reducers::RasterReducer.new(points: seq, options: @options).reduce! if @options[:compress].has_key?(:raster)
          Contrek::Reducers::VisvalingamReducer.new(points: seq, options: @options[:compress][:visvalingam]).reduce! if @options[:compress].has_key?(:visvalingam)
          Contrek::Reducers::DouglasPeuckerReducer.new(points: seq).reduce! if @options[:compress].has_key?(:douglas_peucker)
        end
      end

      # nominal sequence
      def named_sequence
        @plot_sequence.map(&:name).join
      end

      # builds node sequences (with space coordinates) array scanning upper and lower element
      # needs root_nodes ready
      def build_tangs_sequence
        @vert_nodes.each do |line|
          line.each do |node|
            node.precalc_tangs_sequences(cluster: self)
          end
        end
      end

      def plot(bitmap)
        versus = @options[:versus]
        inner_v = VERSUS_INVERTER[versus]
        index_order = 0
        while @root_nodes.size > 0
          root_node = @root_nodes.shift

          root_node.outer_index = index_order
          @plot_sequence = []
          @sequence_coords = []
          bounds = Bounds.empty
          # external polygon
          @plot_sequence << root_node

          next_node = if versus == :a
            root_node.get_tangent_node_by_virtual_index(root_node.tangs_sequence.last)
          else
            root_node.get_tangent_node_by_virtual_index(root_node.tangs_sequence.first)
          end

          if next_node
            beginning_point = (versus == :a) ? root_node.nw_point : root_node.ne_point
            @sequence_coords << beginning_point
            bounds.expand(**beginning_point)

            coord = next_node.coords_entering_to(root_node, VERSUS_INVERTER[versus], Contrek::Finder::Node::OUTER)
            @sequence_coords << coord
            bounds.expand(**coord)
          end

          plot_node(next_node, root_node, bounds, versus) if @nodes > 0 && !next_node.nil?

          if next_node.nil?
            if versus == :a
              nw_point = root_node.nw_point
              @sequence_coords << nw_point
              bounds.expand(**nw_point)
              @sequence_coords << root_node.sw_point
              se_point = root_node.se_point
              @sequence_coords << se_point
              bounds.expand(**se_point)
              @sequence_coords << root_node.ne_point
            else
              ne_point = root_node.ne_point
              @sequence_coords << ne_point
              bounds.expand(**ne_point)
              @sequence_coords << root_node.se_point
              sw_point = root_node.sw_point
              @sequence_coords << sw_point
              bounds.expand(**sw_point)
              @sequence_coords << root_node.nw_point
            end
          end

          draw_sequence(bitmap, "X") unless bitmap.nil?

          if @sequence_coords.any?
            @polygons << {outer: @sequence_coords, inner: [], bounds: (bounds.to_h if @options[:bounds])}.compact
            @sequences << @plot_sequence

            @count = 0
            index_inner = 0
            while @inner_plot.size > 0
              @plot_sequence = []
              @sequence_coords = []
              # mia test
              first = @inner_plot.find { |x| x.tangs_count <= 2 } || @inner_plot.first

              @plot_sequence << first
              @inner_plot.delete(first)
              @root_nodes.delete(first)

              first.inner_index = index_inner

              # @count += 1
              # if @count > 10000
              #  puts "Houston, we have a problem!"
              #  break
              # end

              next_node = if (first.track & Contrek::Finder::Node::OMAX) != 0
                if inner_v == :a
                  vert_nodes[first.y + Node::T_UP][first.upper_start]
                else
                  vert_nodes[first.y + Node::T_DOWN][first.lower_start]
                end
              elsif inner_v == :a
                vert_nodes[first.y + Node::T_DOWN][first.lower_end]
              else
                vert_nodes[first.y + Node::T_UP][first.upper_end]
              end

              if !next_node.nil?
                @sequence_coords << next_node.coords_entering_to(first, inner_v, Contrek::Finder::Node::INNER)
                last_node = plot_inner_node(next_node, inner_v, first, root_node)
                last_coord = last_node.coords_entering_to(first, VERSUS_INVERTER[inner_v], Contrek::Finder::Node::INNER)
                @sequence_coords << last_coord if @sequence_coords.last != last_coord
              end

              draw_sequence(bitmap, "+") unless bitmap.nil?

              @polygons.last[:inner] << @sequence_coords

              @inner_plot.grab(@inner_new)
              index_inner += 1
            end
            # tree
            @treemap << ((versus == :a) ? test_in_hole_a(root_node) : test_in_hole_o(root_node)) if @options.has_key?(:treemap)
            index_order += 1
          end
        end
      end

      def test_in_hole_a(node)
        if node.outer_index > 0
          start_left = node.abs_x_index - 1
          loop do
            prev = @vert_nodes[node.y][start_left]
            if ((cindex = prev.outer_index) < node.outer_index) && ((prev.track & Contrek::Finder::Node::IMAX) != 0)
              start_right = node.abs_x_index
              while (start_right += 1) != @vert_nodes[node.y].size
                tnext = @vert_nodes[node.y][start_right]
                if tnext.outer_index == cindex
                  if (tnext.track & Contrek::Finder::Node::IMIN) != 0
                    return [cindex, (prev.inner_right_index == -1) ? prev.inner_left_index : prev.inner_right_index]
                  else
                    return [-1, -1]
                  end
                end
              end
            end
            break if (start_left -= 1) < 0
          end
        end
        [-1, -1]
      end

      def test_in_hole_o(node)
        if node.outer_index > 0 && @vert_nodes[node.y].last != node
          start_left = node.abs_x_index + 1
          loop do
            prev = @vert_nodes[node.y][start_left]
            if ((cindex = prev.outer_index) < node.outer_index) && ((prev.track & Contrek::Finder::Node::IMIN) != 0)
              start_right = node.abs_x_index
              while (start_right -= 1) >= 0
                tnext = @vert_nodes[node.y][start_right]
                if tnext.outer_index == cindex
                  if (tnext.track & Contrek::Finder::Node::IMAX) != 0
                    return [cindex, (prev.inner_left_index == -1) ? prev.inner_right_index : prev.inner_left_index]
                  else
                    return [-1, -1]
                  end
                end
              end
            end
            break if (start_left += 1) == @vert_nodes[node.y].size
          end
        end
        [-1, -1]
      end

      def draw_sequence(bitmap, val = nil)
        count = 1
        @sequence_coords.each do |coords|
          bitmap.value_set(coords[:x], coords[:y], val.nil? ? count.alph : val)
          count += 1
        end
      end

      # inner way
      # nodes in @plot_sequence
      # coordinates in @sequence_coords
      def plot_inner_node(node, versus, stop_at, start_node)
        node.outer_index = start_node.outer_index
        @root_nodes.delete(node)
        @inner_plot.delete(node)
        last_node = @plot_sequence.last
        next_node = node.my_next(last_node, versus, :inner)
        @plot_sequence << node

        first_is_max = ((node.y > last_node.y) == (versus == :a))
        if first_is_max
          node.inner_right_index = stop_at.inner_index if node.inner_right_index == -1
        elsif node.inner_left_index == -1
          node.inner_left_index = stop_at.inner_index
        end

        plot = true
        if next_node.y == last_node.y
          virtual_index = node.tangs_sequence.send((versus == :a) ? :first : :last)
          plot = (node.get_tangent_node_by_virtual_index(virtual_index) == next_node)
        end

        if plot
          first_point = last_node.coords_entering_to(node, VERSUS_INVERTER[versus], Contrek::Finder::Node::INNER)
          @sequence_coords << first_point if @sequence_coords.last != first_point
          if next_node.y == last_node.y
            if next_node.y < node.y
              pt_a, pt_b = node.sw_point, node.se_point
            else
              pt_a, pt_b = node.ne_point, node.nw_point
            end
            if versus == :o
              @sequence_coords << pt_a
              @sequence_coords << pt_b
            else
              @sequence_coords << pt_b
              @sequence_coords << pt_a
            end
          end
          @sequence_coords << next_node.coords_entering_to(node, versus, Contrek::Finder::Node::INNER)
        end

        if node.track_uncomplete
          @inner_new << node
        else
          @inner_new.delete(node)
        end

        return node if next_node == stop_at
        plot_inner_node(next_node, versus, stop_at, start_node)
      end

      # contour tracing core logic loop
      def plot_node(node, start_node, bounds, versus = :a)
        @root_nodes.delete(node)

        node.outer_index = start_node.outer_index
        last_node = @plot_sequence.last
        next_node = node.my_next(last_node, versus, :outer)

        @plot_sequence << node

        plot = true
        if next_node.y == last_node.y
          virtual_index = node.tangs_sequence.send((versus == :a) ? :last : :first)
          plot = (node.get_tangent_node_by_virtual_index(virtual_index) == next_node)
        end

        if plot
          start_coord = last_node.coords_entering_to(node, versus, Contrek::Finder::Node::OUTER)
          if @sequence_coords.last != start_coord
            @sequence_coords << start_coord
            bounds.expand(**start_coord)
          end

          final_step = node == start_node && node.track_complete
          if next_node.y == last_node.y
            if next_node.y < node.y
              pt_a, pt_b = node.se_point, node.sw_point
            else
              pt_a, pt_b = node.nw_point, node.ne_point
            end
            first = (versus == :o) ? pt_a : pt_b
            second = (versus == :o) ? pt_b : pt_a
            @sequence_coords << first
            bounds.expand(**first)
            unless final_step
              @sequence_coords << second
              bounds.expand(**second)
            end
          end

          unless final_step
            end_coord = next_node.coords_entering_to(node, VERSUS_INVERTER[versus], Contrek::Finder::Node::OUTER)
            @sequence_coords << end_coord
            bounds.expand(**end_coord)
          end

          if node != start_node
            @inner_plot.contains(node) ? @inner_plot.delete(node) : @inner_plot << node
            if last_node.y == next_node.y
              @inner_plot.contains(node) ? @inner_plot.delete(node) : @inner_plot << node
            end
          end
        end
        # exit if root_node
        return if node == start_node && node.track_complete

        plot_node(next_node, start_node, bounds, versus)
      end

      def add_node(node, offset)
        @nodes += 1
        node.abs_x_index = @vert_nodes[node.y].size

        @vert_nodes[node.y] << node
        @root_nodes << node

        if node.y > 0
          # all nodes until up_node.max_x >= node.min_x
          up_nodes = @vert_nodes[node.y - 1]
          up_nodes_count = up_nodes.size
          if up_nodes_count > 0
            index = 0
            loop do
              up_node = up_nodes[index]
              if ((up_node.max_x - 1) + offset) >= node.min_x
                if (up_node.min_x - offset) <= (node.max_x - 1)
                  node.add_intersection(up_node, index)
                  up_node.add_intersection(node, node.abs_x_index)
                end
                return if (index += 1) == up_nodes_count
                loop do
                  up_node = up_nodes[index]
                  if (up_node.min_x - offset) <= (node.max_x - 1)
                    node.add_intersection(up_node, index)
                    up_node.add_intersection(node, node.abs_x_index)
                  else
                    return
                  end
                  break if (index += 1) == up_nodes_count
                end
                return
              end
              break if (index += 1) == up_nodes_count
            end
          end
        end
      end
    end
  end
end
